### PR TITLE
wasmtime: Overhaul trampolines

### PR DESCRIPTION
**Pull request by fitzgen**
This commit splits `VMCallerCheckedFuncRef::func_ptr` into three new function pointers: `VMCallerCheckedFuncRef::{wasm,array,native}_call`. Each one has a dedicated calling convention, so callers just choose the version that works for them. This is as opposed to the previous behavior where we would chain together many trampolines that converted between calling conventions, sometimes up to four on the way into Wasm and four more on the way back out. See [0] for details.

[0] https://github.com/bytecodealliance/rfcs/blob/main/accepted/tail-calls.md#a-review-of-our-existing-trampolines-calling-conventions-and-call-paths

Thanks to @bjorn3 for the initial idea of having multiple function pointers for different calling conventions.

This is generally a nice ~5-10% speed up to our call benchmarks across the board: both Wasm-to-host and host-to-Wasm. The one exception is typed calls from Wasm to the host, which have a minor regression. We hypothesize that this is because the old hand-written assembly trampolines did not maintain a call frame and do a tail call, but the new Cranelift-generated trampolines do maintain a call frame and do a regular call. The regression is only a couple nanoseconds, which seems well-explained by these differences explain, and ultimately is not a big deal.

However, this does lead to a ~5% code size regression for compiled modules. Before, we compiled a trampoline per escaping function's signature and we deduplicated these trampolines by signature. Now we compile two trampolines per escaping function: one for if the host calls via the array calling convention and one for it the host calls via the native calling convention. Additionally, we compile a trampoline for every type in the module, in case there is a native calling convention function from the host that we `call_indirect` of that type. Much of this is in the `.eh_frame` section in the compiled module, because each of our trampolines needs an entry there. Note that the `.eh_frame` section is not required for Wasmtime's correctness, and you can disable its generation to shrink compiled module code size; we just emit it to play nice with external unwinders and profilers. We believe there are code size gains available for follow up work to offset this code size regression in the future.

Backing up a bit: the reason each Wasm module needs to provide these Wasm-to-native trampolines is because `wasmtime::Func::wrap` and friends allow embedders to create functions even when there is no compiler available, so they cannot bring their own trampoline. Instead the Wasm module has to supply it. This in turn means that we need to look up and patch in these Wasm-to-native trampolines during roughly instantiation time. But instantiation is super hot, and we don't want to add more passes over imports or any extra work on this path. So we integrate with `wasmtime::InstancePre` to patch these trampolines in ahead of time.

### CODE REVIEW

**Comment by fitzgen** [comment]
```
Run cargo vet --locked
Vetting Failed!

1 unvetted dependencies:
  bumpalo:3.12.0 missing ["safe-to-deploy"]
```

Unclear why this is failing, since I have a wildcard audit for `bumpalo`.

**Comment by fitzgen** [comment]
I guess I needed to run `cargo vet` without `--locked` locally and commit the changes.

**Comment by fitzgen** [comment]
Everything is passing except s390x -- I wonder if some little endian assumptions snuck in somewhere.

File: crates/environ/src/trap_encoding.rs
Comment (alexcrichton) [inline crates/environ/src/trap_encoding.rs:92]: Should this perhaps be `#[doc(hidden)]`? (this'll end up in the `wasmtime` crate)

Comment (alexcrichton) [reply crates/environ/src/trap_encoding.rs:92]: Or should we perhaps `abort()` if this ever gets seen?

File: crates/jit/src/instantiate.rs
Comment (alexcrichton) [inline crates/jit/src/instantiate.rs:47]: To cut down on the encoding space necessary for this, could these two arrays get fused? And to go a bit further could this perhaps be stored in the `funcs` array directly?

File: crates/jit/src/instantiate.rs
Comment (alexcrichton) [inline crates/jit/src/instantiate.rs:51]: Mind adding a note here about this being sorted as well?

File: crates/jit/src/instantiate.rs
Comment (alexcrichton) [inline crates/jit/src/instantiate.rs:569]: I think the only use case for this is the profilers, and they don't have a great use for this. The profilers already have a lot of duplication of looking at all of these trampolines which isn't great, too.

This is fine to defer to another PR, but I wanted to write this down: these should all get removed and all the profilers should switch to using [this logic](https://github.com/bytecodealliance/wasmtime/blob/43ec4816f1280086987a6743d06368a606406bea/crates/wasmtime/src/trampoline/func.rs#L74-L103) which generically reads an ELF file and registers everything. There's specific support for DWARF in the jitdump profiler but I think that's all best removed at this point.

Again, future PR, just want to leave a reminder to myself to do this or open an issue later.

File: crates/runtime/src/instance.rs
Comment (alexcrichton) [inline crates/runtime/src/instance.rs:518]: TODO

File: crates/runtime/src/trampolines/x86_64.rs
Comment (alexcrichton) [inline crates/runtime/src/trampolines/x86_64.rs:7]: It's probably ok to jettison this now

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [inline crates/runtime/src/vmcontext.rs:23]: Could the inner field here be made private?

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [inline crates/runtime/src/vmcontext.rs:19]: Could this and the two types below have some more documentation about the ABI? E.g. for the structs below a definition of the ABI and for this function pointer a definition of what each of the arguments are?

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [inline crates/runtime/src/vmcontext.rs:50]: Shouldn't this always be `Some(...)` by the time it reaches the `VMFunctionImport` phase?

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [inline crates/runtime/src/vmcontext.rs:641]: The large doc block above about why this can be `None` I think might be good to insert here instead?

File: crates/wasmtime/Cargo.toml
Comment (alexcrichton) [inline crates/wasmtime/Cargo.toml:64]: Could the reformatting here and below be undone?

File: crates/wasmtime/src/component/component.rs
Comment (alexcrichton) [inline crates/wasmtime/src/component/component.rs:1]: Looks like this file became executable?

File: crates/wasmtime/src/func.rs
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:210]: I worry about this because it can accidentally paper over non-send/sync in `kind: FuncKind` and `ty: Option<Box<FuncType>>`. The latter is probably tested elsewhere but there's not a ton of protection here against `FuncKind`.

I don't have a great way to solve this other than unergonomic wrappers, but I'd probably still prefer that to the impl here on the whole struct.

File: crates/wasmtime/src/func.rs
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:1100]: I know we split this into a two-phase operation, but given that the push/fill is always back-to-back it may be better to just remove this `fill_func_refs` perhaps? (folding it into the push operation that is)

File: crates/c-api/include/wasmtime/func.h
Comment (alexcrichton) [reply crates/c-api/include/wasmtime/func.h:248]: Also, could you add documentation about this parameter? Mostly that this isn't "number of arguments" but rather "capacity of the `args_and_results` area" which I realize is perhaps obvious given the name but is not necessarily obvious given that this is generally a "call a function" function where one typically expects to pass a list of arguments.

File: crates/wasmtime/src/instance.rs
Comment (alexcrichton) [inline crates/wasmtime/src/instance.rs:681]: Like with `FuncData`, this feels a bit too broad to me. Could these impls be placed on `VMCallerCheckedFuncRef` instead?

File: crates/wasmtime/src/store/func_refs.rs
Comment (alexcrichton) [inline crates/wasmtime/src/store/func_refs.rs:14]: In retrospect given that this isn't really perf critical or used all that often, perhaps this would be better as a simple `Vec<Box<....>>`?

File: crates/wasmtime/src/store/func_refs.rs
Comment (alexcrichton) [inline crates/wasmtime/src/store/func_refs.rs:30]: Like the above comments, I think this should be removed in favor of something more targeted ideally.

File: crates/winch/src/compiler.rs
Comment (alexcrichton) [inline crates/winch/src/compiler.rs:86]: cc @saulecabrera we opted to leave these as `todo!()` in winch, but this effectively breaks compilation with winch in Wasmtime (and tests are additionally disabled lower in this PR).

Is that ok with you though?

Comment (saulecabrera) [reply crates/winch/src/compiler.rs:86]: That's ok with me, thanks for the heads up. I'll work on the new version of the trampolines for winch once this PR lands.

File: crates/jit/src/instantiate.rs
Comment (fitzgen) [reply crates/jit/src/instantiate.rs:569]: I'll do this in a follow up.

File: crates/runtime/src/trampolines/x86_64.rs
Comment (fitzgen) [reply crates/runtime/src/trampolines/x86_64.rs:7]: Eh, I'd rather just keep it around until we completely remove the asm trampolines.

File: crates/runtime/src/vmcontext.rs
Comment (fitzgen) [reply crates/runtime/src/vmcontext.rs:50]: Oh good call, because the case that the above comment is talking about can't happen with function imports, only `funcref`s.

File: crates/wasmtime/src/func.rs
Comment (fitzgen) [reply crates/wasmtime/src/func.rs:1100]: Ah unfortunately the borrows don't really work out becuase `store` is already borrowed immutably for the `func_refs()` method and so we can't also get and pass in the store's module registry.

File: crates/wasmtime/src/store/func_refs.rs
Comment (fitzgen) [reply crates/wasmtime/src/store/func_refs.rs:14]: Eh, it is still used whenever passing `funcref`s into Wasm via any method other than function imports that are pre-resolved in `InstancePre`. That is a lot of paths, even if it isn't the hottest path. Since we've already got it, I'm inclined to leave it.

File: crates/winch/src/compiler.rs
Comment (fitzgen) [reply crates/winch/src/compiler.rs:86]: Ah yes, sorry I meant to tag you @saulecabrera but I forgot. Thanks for your help getting this PR landed, and willingness to work on follow ups!

**Review by alexcrichton — APPROVED** [review]

**Comment by kpreisser** [comment]
Hi! It seems this change has introduced a regression in the Wasmtime C API, which causes a panic similar to the following to occur e.g. in the [`wasmtime-dotnet` unit tests](https://github.com/bytecodealliance/wasmtime-dotnet/actions/runs/4828116489/jobs/8601557772?pr=245):
```
thread '<unnamed>' panicked at '0x0 > 0xdeb7f7d2f0', crates\runtime\src\traphandlers\backtrace.rs:301:13
```

Please see https://github.com/bytecodealliance/wasmtime-dotnet/pull/245#issuecomment-1527820325 for more details and for a reproduction case using `wasmtime-dotnet` (C#, .NET 7.0).
Thanks!

**Pull request merged**
